---
title: "TCGA_LAML_breakpoint_coverage_profiles.Rmd"
author: "Kristen Dang"
date: "December 10, 2015"
output: html_document
---

Iterates through sample breakpoints, loads coverage files for that sample.


```{r}
library(synapseClient)
library(GenomicRanges)
library(rtracklayer)
library(biomaRt)
library(Gviz)
library(GenomicFeatures)

setwd('/work/DAT_086__TCGA_LAML/Results/window_500')
Hs=useMart("ENSEMBL_MART_ENSEMBL", host="www.ensembl.org") # use this one when biomart.org is down
Hs = useDataset("hsapiens_gene_ensembl", Hs)
ENSG_TxDb  = makeTxDbFromGFF('/external-data/Genome/gene_models/Hsapiens_Ensembl_v82_primary.gtf',format=c("gtf"),dataSource='Ensembl',organism='Homo sapiens')


synapseLogin()
```

Get breakpoint list and lists of all bedgraph files
```{r}
breakpointEnt = synGet("syn5548948")
breakpoints = read.delim(getFileLocation(breakpointEnt), header = FALSE, skip = 1)
colnames(breakpoints) = c("chr", "start", "end", "sample")
head(breakpoints)


i = 1
breakpoints[i,]

tmp = dir('/work/DAT_086__TCGA_LAML/Data')
allBedgraphs = tmp[grep(pattern="bedgraph", tmp)]
head(allBedgraphs)
chimericBedgraphs = tmp[grep(pattern="Chimeric.out.bedgraph", tmp)]
head(chimericBedgraphs)
notChimericBedgraphs = tmp[grep(pattern="Coord.out.bedgraph", tmp)]
head(notChimericBedgraphs)
```


```{r}
idKeyEnt = synGet("syn5548599")
idKey = read.delim(getFileLocation(idKeyEnt))
idKey$mainID = substr(idKey$barcode,start = 9, stop = 12)
head(idKey)

even = seq(2,nrow(breakpoints),by=2)
odd = seq(1,nrow(breakpoints),by=2)

window = 500
for (i in 1:nrow(breakpoints)){
  #  i = 1
  dataFileNames = as.character(idKey$analysis_id[which(idKey$mainID ==breakpoints$sample[i])])
  for (j in 1:length(dataFileNames)){
    #    j = 1
    #    toExtract = grep(pattern=dataFileNames[j],x=allBedgraphs)
    #  track = import(paste("/work/DAT_086__TCGA_LAML/Data", allBedgraphs[toExtract[1]],sep = "/"),format="bedgraph")
    
    bgFile = paste("/work/DAT_086__TCGA_LAML/Data/", dataFileNames[j], "Chimeric.out.bedgraph", sep = "")
    #bgFile = paste("/work/DAT_086__TCGA_LAML/Data/", dataFileNames[j], "Aligned.sortedByCoord.out.bedgraph", sep = "")
    if (!file.exists(bgFile)) { next }
    donorTitle = paste('chimeric',breakpoints$sample[i],breakpoints$chr[i],breakpoints$chr[i-1],"donor", i, sep = "-")
    acceptorTitle = paste('chimeric',breakpoints$sample[i],breakpoints$chr[i+1],breakpoints$chr[i],"acceptor",i,  sep = "-")
    
    dTrack = DataTrack(range = bgFile, genome = "hg38",type = "l", chromosome = breakpoints$chr[i], start = breakpoints$start[i], end = breakpoints$end[i], name = "bedGraph", options(ucscChromosomeNames=FALSE))
    grtrack = GeneRegionTrack(ENSG_TxDb, chromosome = breakpoints$chr[i], start = breakpoints$start[i], end = breakpoints$end[i], options(ucscChromosomeNames=FALSE))
    gtrack <- GenomeAxisTrack()
    if (i %in% odd) { # First listed, but actually the acceptor, use start + window size
      pdf(file=paste(acceptorTitle, "pdf", sep = "."))
      plotTracks(list(dTrack, gtrack, grtrack), transcriptAnnotation = "symbol", from = breakpoints$start[i]-window, to = breakpoints$start[i]+window, main = acceptorTitle)
      dev.off()
      }
    else { # Second listed, but actually the donor, use end - window size
      pdf(file=paste(donorTitle, "pdf", sep = "."))
      plotTracks(list(dTrack, gtrack, grtrack), transcriptAnnotation = "symbol", from = breakpoints$end[i]-window, to = breakpoints$end[i]+window, main = donorTitle)
      dev.off()
      }
    }
  }
```